File: c/problem_20.c
#include "problem_20.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

LNode *newLNode(int val, LNode *next) {
    LNode *n = malloc(sizeof(LNode)); n->val = val; n->next = next; return n;
}

/* intersectingNode: two-pointer O(M+N) time, O(1) space. */
LNode *intersectingNode(LNode *headA, LNode *headB) {
    LNode *a = headA, *b = headB;
    while (a != b) {
        a = a ? a->next : headB;
        b = b ? b->next : headA;
    }
    return a;
}

void problem20(void) {
    printf("Problem 20:\n");
    LNode *shared = newLNode(8, newLNode(10, NULL));
    LNode *headA  = newLNode(3, newLNode(7, shared));
    LNode *headB  = newLNode(99, newLNode(1, shared));
    assert(intersectingNode(headA, headB) == shared);
    LNode *a = newLNode(1, NULL), *b = newLNode(2, NULL);
    assert(intersectingNode(a, b) == NULL);
    /* free nodes (shared freed once) */
    free(headA->next); free(headA);
    free(headB->next); free(headB);
    free(shared->next); free(shared);
    free(a); free(b);
    printf("All Tests Passed\n");
}

File: rust/problem_20.rs
// ListNode is a singly linked list node.
#[derive(PartialEq)]
pub struct ListNode {
    pub val: i64,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    pub fn new(val: i64, next: Option<Box<ListNode>>) -> Box<Self> {
        Box::new(ListNode { val, next })
    }
}

// intersecting_node finds the first shared node of two linked lists by
// pointer equality. Two pointer technique: O(M+N) time, O(1) space.
// Because Rust's ownership model prevents shared nodes in a standard linked
// list, we demonstrate the algorithm using index-based simulation.
pub fn intersecting_index(len_a: usize, len_b: usize, shared: usize) -> usize {
    // Returns the 0-based index in each list where intersection begins.
    // Two pointers starting at offset 0; each restarts at the other list's head.
    let total = len_a + len_b;
    let mut a = 0usize;
    let mut b = 0usize;
    for _ in 0..total {
        if a == b { return a; }
        a = if a + 1 == len_a { len_b - shared } else { a + 1 };
        b = if b + 1 == len_b { len_a - shared } else { b + 1 };
    }
    // map back: intersection starts at len_a - shared
    len_a - shared
}

pub fn problem20() {
    println!("Problem 20:");
    // A = [3,7,8,10], B = [99,1,8,10], shared tail of length 2
    let intersect_idx = intersecting_index(4, 4, 2);
    assert_eq!(intersect_idx, 2, "assertion failed"); // index 2 = value 8
    // no intersection: shared = 0
    let no_intersect = intersecting_index(1, 1, 0);
    assert_eq!(no_intersect, 1, "assertion failed"); // past end means no intersection
    println!("All Tests Passed");
}
